internal/middleware/enforce_identity.go

```diff
@@ -0,0 +1,89 @@
+package middleware
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/redhatinsights/platform-go-middlewares/identity"
+)
+
+var (
+	InvalidOrMissingOrgErr = errors.New("x-rh-identity header has an invalid or missing org_id")
+	MissingTypeErr         = errors.New("x-rh-identity header is missing type")
+)
+
+func doError(ctx context.Context, w http.ResponseWriter, code int, reason string) {
+	logger := ctxval.Logger(ctx)
+	logger.Warn().Msgf("Failed to enforce the Identity header: %s", reason)
+	http.Error(w, reason, code)
+}
+
+// EnforceIdentity extracts the X-Rh-Identity header and places the contents into the
+// request context.  If the Identity is invalid, the request will be aborted.
+func EnforceIdentity(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger := ctxval.Logger(r.Context())
+		rawHeaders := r.Header["X-Rh-Identity"]
+
+		// must have an x-rh-id header
+		if len(rawHeaders) != 1 {
+			doError(r.Context(), w, 400, "missing x-rh-identity header")
+			return
+		}
+
+		// must be able to base64 decode header
+		idRaw, err := base64.StdEncoding.DecodeString(rawHeaders[0])
+		if err != nil {
+			doError(r.Context(), w, 400, "unable to b64 decode x-rh-identity header")
+			return
+		}
+
+		var jsonData identity.XRHID
+		err = json.Unmarshal(idRaw, &jsonData)
+		if err != nil {
+			logger.Warn().Err(err).Msg("unable to unmarshal x-rh-identity header")
+			doError(r.Context(), w, 400, "x-rh-identity header does not contain valid JSON")
+			return
+		}
+
+		topLevelOrgIDFallback(&jsonData)
+
+		err = checkHeader(r.Context(), &jsonData, w)
+		if err != nil {
+			return
+		}
+
+		ctx := context.WithValue(r.Context(), identity.Key, jsonData)
+		next.ServeHTTP(w, r.WithContext(ctx))
```

Comment:
Can we use our own `ctxval` setter and getter with our own key then? For consistency, I never liked that :-)

Reply:
We would be polluting more code with this... but I guess in case it stays...

Reply:
btw we should not use our own key. some other services might want to share the header, I'd keep it under globally shared key, just in case there is more shared code?

Reply:
If you'd not mind, I'd not do this, I'd love to keep it as much unchanged from the original as possible.

Reply:
This will stay forever, trust me. We will never touch this again if it works. I can do it later tho.